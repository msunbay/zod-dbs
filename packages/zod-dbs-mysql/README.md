# zod-dbs-mysql

MySQL provider for [zod-dbs](https://github.com/msunbay/zod-dbs).

## Installation

```bash
npm install zod-dbs-mysql
```

> Note: This package has a peer dependency on `mysql2`. If you don't already have it, install it:

```bash
npm install mysql2
```

## Cli Usage

```bash
npm install zod-dbs-cli zod-dbs-mysql

npx zod-dbs --provider mysql
```

## Programmatic Usage

```ts
import { generateZodSchemas } from 'zod-dbs';
import { MySqlProvider } from 'zod-dbs-mysql';

const provider = new MySqlProvider();

await generateZodSchemas({
  provider,
  config: {
    // See zod-dbs documentation for available options
  },
});
```

## Compatibility

- Works with MariaDB in addition to MySQL. The provider uses the MySQL protocol (`mysql2`) and introspects via `information_schema.columns`, which MariaDB supports.
