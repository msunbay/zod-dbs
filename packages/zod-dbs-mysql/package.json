{
  "name": "zod-dbs-mysql",
  "version": "1.0.0",
  "description": "MySQL provider for zod-dbs",
  "main": "./dist/index.js",
  "type": "module",
  "scripts": {
    "build": "rm -rf ./dist && tsc --project tsconfig.build.json",
    "test:unit": "vitest --project unit run",
    "test:integration": "vitest --project integration run",
    "test:types": "tsc --noEmit"
  },
  "author": "msolvaag",
  "license": "ISC",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/msolvaag/zod-dbs.git",
    "directory": "packages/zod-dbs-mysql"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "dist"
  ],
  "dependencies": {
    "zod-dbs-core": "workspace:*"
  },
  "devDependencies": {
    "@testcontainers/mysql": "^11.5.1",
    "mysql2": "^3.14.3"
  },
  "peerDependencies": {
    "mysql2": ">=2"
  }
}
