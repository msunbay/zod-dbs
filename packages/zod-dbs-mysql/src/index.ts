export { createClient } from './client.js';
export { MySqlProvider } from './MySqlProvider.js';
export { createProvider } from './provider.js';
export type * from './types.js';
