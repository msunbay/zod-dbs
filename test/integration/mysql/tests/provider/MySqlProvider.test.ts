import { MySqlProvider } from 'zod-dbs-mysql';

import { getProviderConfig } from '../../../utils/context.js';

const connectionConfig = getProviderConfig();

it('returns raw schema column information', async () => {
  const provider = new MySqlProvider();

  const info = await provider.fetchSchemaInfo({
    ...connectionConfig,
    schemaName: connectionConfig.database!,
  });

  expect(info).toBeDefined();
  expect(info).toHaveLength(7);
  expect(info).toMatchSnapshot('rawColumns');
});

it('returns schema models', async () => {
  const provider = new MySqlProvider();

  const info = await provider.getSchemaInformation(connectionConfig);

  expect(info).toBeDefined();
  expect(info.tables).toBeDefined();
  expect(info.tables).toHaveLength(1);

  const userTable = info.tables.find((t) => t.name === 'users')!;
  expect(userTable).toBeDefined();
  expect(userTable.columns).toBeDefined();
  expect(userTable.columns).toHaveLength(7);

  expect(userTable).toMatchSnapshot('userTable');
});

it('throws if schemaName and database is missing', async () => {
  const provider = new MySqlProvider();

  await expect(
    provider.fetchSchemaInfo({
      ...connectionConfig,
      database: undefined,
      schemaName: undefined,
    })
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `[Error: MySQL provider requires a database/schema name to fetch schema information.]`
  );
});
