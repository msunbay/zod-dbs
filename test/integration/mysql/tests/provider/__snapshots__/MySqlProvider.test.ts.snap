// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns raw schema column information > rawColumns 1`] = `
[
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": "",
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": true,
    "maxLen": undefined,
    "name": "id",
    "schemaName": "test",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "varchar",
    "defaultValue": undefined,
    "description": "",
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": 100,
    "name": "name",
    "schemaName": "test",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "varchar",
    "defaultValue": undefined,
    "description": "",
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "maxLen": 255,
    "name": "email",
    "schemaName": "test",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "varchar",
    "defaultValue": "active",
    "description": "",
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "maxLen": 20,
    "name": "status",
    "schemaName": "test",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "json",
    "defaultValue": undefined,
    "description": "",
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "maxLen": undefined,
    "name": "profile",
    "schemaName": "test",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "timestamp",
    "defaultValue": "CURRENT_TIMESTAMP",
    "description": "",
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "maxLen": undefined,
    "name": "created_at",
    "schemaName": "test",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "enum",
    "defaultValue": "viewer",
    "description": "",
    "enumValues": [
      "admin",
      "editor",
      "viewer",
    ],
    "isArray": false,
    "isEnum": true,
    "isNullable": true,
    "isSerial": false,
    "maxLen": 6,
    "name": "roles",
    "schemaName": "test",
    "tableName": "users",
    "tableType": "table",
  },
]
`;

exports[`returns schema models > userTable 1`] = `
{
  "columns": [
    {
      "dataType": "int",
      "defaultValue": undefined,
      "description": "",
      "isArray": false,
      "isEnum": false,
      "isNullable": false,
      "isReadOptional": false,
      "isSerial": true,
      "isWritable": false,
      "isWriteOptional": false,
      "maxLen": undefined,
      "name": "id",
      "schemaName": "test",
      "tableName": "users",
      "tableType": "table",
      "type": "int",
    },
    {
      "dataType": "varchar",
      "defaultValue": undefined,
      "description": "",
      "isArray": false,
      "isEnum": false,
      "isNullable": false,
      "isReadOptional": false,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": false,
      "maxLen": 100,
      "name": "name",
      "schemaName": "test",
      "tableName": "users",
      "tableType": "table",
      "type": "string",
    },
    {
      "dataType": "varchar",
      "defaultValue": undefined,
      "description": "",
      "isArray": false,
      "isEnum": false,
      "isNullable": true,
      "isReadOptional": true,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": true,
      "maxLen": 255,
      "name": "email",
      "schemaName": "test",
      "tableName": "users",
      "tableType": "table",
      "type": "string",
    },
    {
      "dataType": "varchar",
      "defaultValue": "active",
      "description": "",
      "isArray": false,
      "isEnum": false,
      "isNullable": true,
      "isReadOptional": true,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": true,
      "maxLen": 20,
      "name": "status",
      "schemaName": "test",
      "tableName": "users",
      "tableType": "table",
      "type": "string",
    },
    {
      "dataType": "json",
      "defaultValue": undefined,
      "description": "",
      "isArray": false,
      "isEnum": false,
      "isNullable": true,
      "isReadOptional": true,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": true,
      "maxLen": undefined,
      "name": "profile",
      "schemaName": "test",
      "tableName": "users",
      "tableType": "table",
      "type": "json",
    },
    {
      "dataType": "timestamp",
      "defaultValue": "CURRENT_TIMESTAMP",
      "description": "",
      "isArray": false,
      "isEnum": false,
      "isNullable": true,
      "isReadOptional": true,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": true,
      "maxLen": undefined,
      "name": "created_at",
      "schemaName": "test",
      "tableName": "users",
      "tableType": "table",
      "type": "date",
    },
    {
      "dataType": "enum",
      "defaultValue": "viewer",
      "description": "",
      "enumValues": [
        "admin",
        "editor",
        "viewer",
      ],
      "isArray": false,
      "isEnum": true,
      "isNullable": true,
      "isReadOptional": true,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": true,
      "maxLen": 6,
      "name": "roles",
      "schemaName": "test",
      "tableName": "users",
      "tableType": "table",
      "type": "any",
    },
  ],
  "name": "users",
  "schemaName": "test",
  "type": "table",
}
`;
